components/search/src/filter.rs

```diff
@@ -158,12 +160,52 @@ fn apply_overrides(engines: &mut [SearchEngineDefinition], overrides: &[JSONOver
     }
 }
 
+fn negotiate_languages(user_environment: &mut SearchUserEnvironment, available_locales: &[String]) {
+    let user_locale = user_environment.locale.to_lowercase();
+
+    let locales_map_to_en_us: HashSet<String> =
+        ENGLISH_LOCALES.iter().map(|s| s.to_lowercase()).collect();
+
+    let available_locales_set: HashSet<String> = available_locales
+        .iter()
+        .map(|locale| locale.to_lowercase())
+        .collect();
+
+    if available_locales_set.contains(&user_locale) {
+        return;
+    }
+    if locales_map_to_en_us.contains(&user_locale) {
+        user_environment.locale = "en-us".to_string();
+        return;
+    }
+    if let Some(index) = user_locale.find('-') {
+        let base_locale = &user_locale[..index];
+        if available_locales_set.contains(base_locale) {
+            user_environment.locale = base_locale.to_string();
+        }
+    }
+}
+
 impl Filter for Vec<RemoteSettingsRecord> {
     fn filter_records(
         &self,
         user_environment: &SearchUserEnvironment,
         overrides: Option<Vec<JSONOverridesRecord>>,
     ) -> Result<FilterRecordsResult, Error> {
+        let mut available_locales = Vec::new();
+        for record in self {
+            let stringified = serde_json::to_string(&record.fields)?;
+            if let Some(val) = record.fields.get("recordType") {
+                if *val == "availableLocales" {
+                    let locales_record: Option<JSONAvailableLocalesRecord> =
+                        serde_json::from_str(&stringified)?;
```

Comment:
Please can you put the `serde_json::to_string` statement within the second if statement to avoid it being called unnecessarily.